File: Java/logger/Logger.java
package logger;

import logger.enums.LoggerColor;
import logger.enums.LoggerOption;
import logger.enums.LoggerType;

import java.io.*;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.Arrays;

import static logger.enums.LoggerOption.*;
import static logger.enums.LoggerType.*;

/*
 * Logger
 *
 * Use Logger.init() to start the logger and Logger.exit() to close the logger
 *
 * Simple usage :
 * Logger.debug("Is it simple ? YES");
 * Write 'Is it simple ? YES' (without the quote) in the console and the file
 *
 * Complex usage :
 * Logger.info("Not too complex ?", "Maybe", LoggerOption.CONSOLE_ONLY);
 * Write 'Not toot complex ? Maybe' (without the quote) only in the console
 */
public abstract class Logger {
    /**
     * The log directory's path
     * Change it with your path
     */
    public static String logPath = "./logs";
    /**
     * Project's name
     * Change it with your project's name
     */
    public static String projectName = "project";

    // _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

    /*
     * Different rules for the formats :
     * %Y -> Year
     * %M -> Month
     * %D -> Day
     * %H -> Hour
     * %m -> Minute
     * %S -> Second
     * %N -> Nano second
     * %d -> Date (%Y-%M-%D@%H-%m-%S)
     * %h -> Hour (%H:%m:%S:%N)
     * %T -> Trace
     * %C -> Content message
     * %n -> Log number
     * %t -> Log type
     */

    /**
     * Format for console log
     */
    public static String console_format = "[%T]\t%C";
    /**
     * Format for log file
     */
    public static String file_format = "[%n-%h-%t]\t[%T]\t%C";
    /**
     * Format for additional output streams
     */
    public static String additional_format = "[%n-%t]\t[%T]\t%C";

    // _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

    /**
     * If the logger is initialized
     */
    private static boolean isInitialized = false;
    /**
     * The writer for the file
     */
    private static PrintWriter printWriter = null;
    /**
     * Additional output for the logs
     */
    private static ArrayList<OutputStream> additionalStreams;
    /**
     * The number of log
     */
    private static int nbLog = 0;
    /**
     * The type of verbose
     */
    private static LoggerOption verbose;
    /**
     * The types of logs that be shown
     */
    private static ArrayList<LoggerType> showTypes;

    // _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

    public static void init() {
        init(FILE_AND_CONSOLE, new LoggerType[]{INFO, SUCCESS, ERROR, WARNING, DEBUG});
    }

    public static void init(LoggerOption verbose) {
        init(verbose, new LoggerType[]{INFO, SUCCESS, ERROR, WARNING, DEBUG});
    }

    public static void init(LoggerType[] showTypes) {
        init(FILE_AND_CONSOLE, showTypes);
    }

    /**
     * Initialisation
     *
     * @param verbose   LoggerOption
     * @param showTypes LoggerType
     */
    public static void init(LoggerOption verbose, LoggerType[] showTypes) {
        if (!isInitialized) {
            Logger.verbose = verbose;
            Logger.showTypes = new ArrayList<>(Arrays.asList(showTypes));
            additionalStreams = new ArrayList<>();

            boolean ok = true;
            boolean dirCreated = false;

            if (verbose != CONSOLE_ONLY) {
                BufferedWriter bufferedWriter;

                try {
                    dirCreated = new File(logPath).mkdir();
                    FileWriter fileWriter = new FileWriter(logPath + "/" + projectName + "_log_" + getDate() + ".log");
                    bufferedWriter = new BufferedWriter(fileWriter);
                    printWriter = new PrintWriter(bufferedWriter);
                } catch (IOException e) {
                    error(e, CONSOLE_ONLY);
                    ok = false;
                }
            }

            if (ok) {
                isInitialized = true;
                info("Log start", FILE_ONLY);

                if (dirCreated) {
                    warning("Log directory created");
                }
            } else {
                error("Log error", CONSOLE_ONLY);
            }
        } else {
            warning("Log already init");
        }
    }

    // _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

    /**
     * Add a new output for the logs
     *
     * @param os OutputStream
     */
    public static void addOutputStream(OutputStream os) {
        additionalStreams.add(os);
    }

    // _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

    /**
     * Generic log use for all logs
     *
     * @param args Object[]
     * @param type LoggerType
     */
    private synchronized static void genericLog(Object[] args, LoggerType type) {
        ArrayList<String> messages = new ArrayList<>();
        ArrayList<LoggerOption> options = new ArrayList<>();

        for (Object arg : args) {
            if (arg instanceof LoggerOption) {
                options.add((LoggerOption) arg);
            } else {
                messages.add(String.valueOf(arg));
            }
        }

        final String separator = " ";

        StringBuilder message = new StringBuilder();

        for (String s : messages) {
            message.append(s);
            message.append(separator);
        }

        StackTraceElement[] traces = Thread.currentThread().getStackTrace();
        String trace = "";
        if (traces.length > 3) {
            StackTraceElement ste = traces[3];
            trace = ste.getClassName() + "." + ste.getMethodName();
        }

        if (!options.contains(FILE_ONLY) && verbose != FILE_ONLY && showTypes.contains(type)) {
            System.out.println(type.getColor() + constructMessage(message.toString(), trace, type.toString(), console_format) + LoggerColor.DEFAULT);
        }

        if (!options.contains(CONSOLE_ONLY) && verbose != CONSOLE_ONLY) {
            writeToFile(constructMessage(message.toString(), trace, type.toString(), file_format));
        }

        if (!options.contains(FILE_ONLY) && !options.contains(CONSOLE_ONLY) && verbose == FILE_AND_CONSOLE) {
            for (OutputStream additionalStream : additionalStreams) {
                try {
                    additionalStream.write((constructMessage(message.toString(), trace, type.toString(), additional_format) + "\n").getBytes());
                    additionalStream.flush();
                } catch (IOException e) {
                    error(e, CONSOLE_ONLY);
                }
            }
        }

        nbLog++;
    }

    /**
     * Construct the message from the format
     * <p>
     * Different rules for the formats :
     * %Y -> Year
     * %M -> Month
     * %D -> Day
     * %H -> Hour
     * %m -> Minute
     * %S -> Second
     * %N -> Nano second
     * %d -> Date (%Y-%M-%D@%H-%m-%S)
     * %h -> Hour (%H:%m:%S:%N)
     * %T -> Trace
     * %C -> Content message
     * %n -> Log number
     * %t -> Log type
     *
     * @param message String
     * @param trace   String
     * @param logType String
     * @param format  String
     * @return String
     */
    private static String constructMessage(String message, String trace, String logType, String format) {
        StringBuilder res = new StringBuilder();

        int i = 0;
        while (i < format.length()) {
            char c = format.charAt(i);
            if (c == '%') {
                i++;
                c = format.charAt(i);

                LocalDateTime now = LocalDateTime.now();
                switch (c) {
                    case 'Y':
                        res.append(now.getYear());
                        break;
                    case 'M':
                        res.append(String.format("%02d", now.getMonthValue()));
                        break;
                    case 'D':
                        res.append(String.format("%02d", now.getDayOfMonth()));
                        break;
                    case 'H':
                        res.append(String.format("%02d", now.getHour()));
                        break;
                    case 'm':
                        res.append(String.format("%02d", now.getMinute()));
                        break;
                    case 'S':
                        res.append(String.format("%02d", now.getSecond()));
                        break;
                    case 'N':
                        res.append(String.format("%03d", now.getNano() / 1000));
                        break;
                    case 'd':
                        res.append(getDate());
                        break;
                    case 'h':
                        res.append(getHour());
                        break;
                    case 'T':
                        res.append(trace);
                        break;
                    case 'C':
                        res.append(message);
                        break;
                    case 'n':
                        res.append(nbLog);
                        break;
                    case 't':
                        res.append(logType);
                        break;
                }
            } else {
                res.append(c);
            }

            i++;
        }

        return res.toString();
    }

    /**
     * Info
     *
     * @param args Object...
     */
    public static void info(Object... args) {
        genericLog(args, INFO);
    }

    /**
     * Success
     *
     * @param args Object...
     */
    public static void success(Object... args) {
        genericLog(args, SUCCESS);
    }

    /**
     * Error
     *
     * @param args Object...
     */
    public static void error(Object... args) {
        genericLog(args, ERROR);
    }

    /**
     * Warning
     *
     * @param args Object...
     */
    public static void warning(Object... args) {
        genericLog(args, WARNING);
    }

    /**
     * Debug
     *
     * @param args Object...
     */
    public static void debug(Object... args) {
        genericLog(args, DEBUG);
    }

    // _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

    /**
     * Write the log into the file
     *
     * @param message String
     */
    private synchronized static void writeToFile(String message) {
        if (printWriter != null) {
            printWriter.println(message);
            printWriter.flush();
        } else if (!isInitialized) {
            error("Please init Logger", CONSOLE_ONLY);
        }
    }

    // _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

    /**
     * The log's hour
     * hh:mm:ss:nnn
     *
     * @return String
     */
    private static String getHour() {
        LocalTime now = LocalTime.now();

        return String.format("%02d", now.getHour())
                + ":" + String.format("%02d", now.getMinute())
                + ":" + String.format("%02d", now.getSecond())
                + ":" + String.format("%03d", now.getNano() / 1000);
    }

    /**
     * The log's date
     * yyyy-mm-dd@hh-mm-ss
     *
     * @return String
     */
    private static String getDate() {
        LocalDateTime now = LocalDateTime.now();

        return now.getYear()
                + "-" + String.format("%02d", now.getMonthValue())
                + "-" + String.format("%02d", now.getDayOfMonth())
                + "@" + String.format("%02d", now.getHour())
                + "-" + String.format("%02d", now.getMinute())
                + "-" + String.format("%02d", now.getSecond());
    }

    // _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

    /**
     * Quit the log and close the writer
     */
    public static void exit() {
        if (isInitialized) {
            info("End log", FILE_ONLY);
            isInitialized = false;

            if (printWriter != null) {
                printWriter.close();
                printWriter = null;
            }
        } else {
            error("Please init before exit", CONSOLE_ONLY);
        }
    }
}

File: Kotlin/logger/Logger.kt
package logger

import logger.enums.LoggerColor
import logger.enums.LoggerOption
import logger.enums.LoggerOption.*
import logger.enums.LoggerType
import logger.enums.LoggerType.*
import java.io.*
import java.time.LocalDateTime

/**
 * Logger
 * <p>
 * Use init() to start the logger and exit() to close the logger
 * <p>
 * Simple usage :
 * debug("Is it simple ? YES")
 * Write 'Is it simple ? YES' (without the quote) in the console and the file
 * <p>
 * Complex usage :
 * info("Not too complex ?", "Maybe", LoggerOption.CONSOLE_ONLY)
 * Write 'Not toot complex ? Maybe' (without the quote) only in the console
 */

/**
 * The log directory's path
 * Change it with your path
 */
internal const val logPath = "./logs"

/**
 * Project's name
 * Change it with your project's name
 */
internal const val projectName = "project"

// _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

/*
 * Different rules for the formats :
 * %Y -> Year
 * %M -> Month
 * %D -> Day
 * %H -> Hour
 * %m -> Minute
 * %S -> Second
 * %N -> Nano second
 * %d -> Date (%Y-%M-%D@%H-%m-%S)
 * %h -> Hour (%H:%m:%S:%N)
 * %T -> Trace
 * %C -> Content message
 * %n -> Log number
 * %t -> Log type
 */

/**
 * Format for console log
 */
internal const val console_format = "[%T]\t%C"

/**
 * Format for log file
 */
internal const val file_format = "[%n-%h-%t]\t[%T]\t%C"

/**
 * Format for additional output streams
 */
internal const val additional_format = "[%n-%t]\t[%T]\t%C"

// _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

/**
 * The writer for the file
 */
internal var printWriter: PrintWriter? = null

/**
 * Additional output for the logs
 */
internal var additionalStreams = ArrayList<OutputStream>()

/**
 * The number of log
 */
internal var nbLog = 0

/**
 * The type of verbose
 */
internal var verbose = FILE_AND_CONSOLE

/**
 * The types of logs that be shown
 */
internal var showTypes = ArrayList<LoggerType>()

// _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

/**
 * Initialisation
 */
fun init(
    verboseP: LoggerOption = FILE_AND_CONSOLE,
    showTypesP: ArrayList<LoggerType> = arrayListOf(INFO, SUCCESS, ERROR, WARNING, DEBUG)
) {
    if (printWriter == null) {
        verbose = verboseP
        showTypes = showTypesP

        var dirCreated = false
        var bufferedWriter: BufferedWriter? = null

        val ok = try {
            dirCreated = File(logPath).mkdir()
            val fileWriter = FileWriter(logPath + "/" + projectName + "_log_" + getDate() + ".log")
            bufferedWriter = BufferedWriter(fileWriter)

            true
        } catch (e: IOException) {
            error(e, CONSOLE_ONLY)
            false
        }

        if (ok) {
            printWriter = PrintWriter(bufferedWriter)

            info("Log start", FILE_ONLY)

            if (dirCreated)
                warning("Log directory created")
        } else
            error("Log error")
    } else
        warning("Log already init")
}

/**
 * Quit the log and close the file
 */
fun exit() {
    if (printWriter != null) {
        info("End log", FILE_ONLY)
        printWriter!!.close()
        printWriter = null
    } else
        error("Please init before exit", CONSOLE_ONLY)
}

// _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

/**
 * Add a new output for the logs
 */
fun addOutputStream(os: OutputStream) {
    additionalStreams.add(os)
}

// _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

/**
 * Generic log use for all logs
 */
internal fun genericLog(args: Array<out Any>, type: LoggerType) {
    val messages = ArrayList<String>()
    val options = ArrayList<LoggerOption>()

    for (arg in args) {
        if (arg is LoggerOption)
            options.add(arg)
        else
            messages.add(arg.toString())
    }

    val separator = " "

    val message = StringBuilder()

    for (i in 0 until messages.size) {
        message.append(messages[i])

        if (i < messages.size)
            message.append(separator)
    }

    val traces = Thread.currentThread().stackTrace
    var trace = ""
    if (traces.size > 3) {
        val ste = traces[3]
        trace = ste.className + "." + ste.methodName
    }

    if (!options.contains(FILE_ONLY) && verbose != FILE_ONLY && showTypes.contains(type))
        println(
            type.color.toString() + constructMessage(
                message.toString(),
                trace,
                type.toString(),
                console_format
            ) + LoggerColor.DEFAULT
        )

    if (!options.contains(CONSOLE_ONLY) && verbose != CONSOLE_ONLY)
        writeToFile(constructMessage(message.toString(), trace, type.toString(), file_format))

    if (!options.contains(FILE_ONLY) && !options.contains(CONSOLE_ONLY) && verbose == FILE_AND_CONSOLE) {
        additionalStreams.forEach {
            try {
                it.write((constructMessage(message.toString(), trace, type.toString(), additional_format) + "\n").toByteArray())
                it.flush()
            } catch (e: IOException) {
                error(e, CONSOLE_ONLY)
            }
        }
    }

    nbLog++
}

/**
 * Construct the message from the format
 * Different rules for the formats :
 * %Y -> Year
 * %M -> Month
 * %D -> Day
 * %H -> Hour
 * %m -> Minute
 * %S -> Second
 * %N -> Nano second
 * %d -> Date (%Y-%M-%D@%H-%m-%S)
 * %h -> Hour (%H:%m:%S:%N)
 * %T -> Trace
 * %C -> Content message
 * %n -> Log number
 * %t -> Log type
 */
internal fun constructMessage(message: String, trace: String, logType: String, format: String): String {
    val res = StringBuilder()

    var i = 0
    while (i < format.length) {
        var c = format[i]
        if (c == '%') {
            i++
            c = format[i]

            val now = LocalDateTime.now()
            when (c) {
                'Y' -> res.append(now.year)
                'M' -> res.append(String.format("%02d", now.monthValue))
                'D' -> res.append(String.format("%02d", now.dayOfMonth))
                'H' -> res.append(String.format("%02d", now.hour))
                'm' -> res.append(String.format("%02d", now.minute))
                'S' -> res.append(String.format("%02d", now.second))
                'N' -> res.append(String.format("%03d", now.nano / 1000))
                'd' -> res.append(getDate())
                'h' -> res.append(getHour())
                'T' -> res.append(trace)
                'C' -> res.append(message)
                'n' -> res.append(nbLog)
                't' -> res.append(logType)
            }
        } else {
            res.append(c)
        }

        i++
    }

    return res.toString()
}

/**
 * Info
 */
fun info(vararg args: Any) = genericLog(args, INFO)

/**
 * Success
 */
fun success(vararg args: Any) = genericLog(args, SUCCESS)

/**
 * Error
 */
fun error(vararg args: Any) = genericLog(args, ERROR)

/**
 * Warning
 */
fun warning(vararg args: Any) = genericLog(args, WARNING)

/**
 * Debug
 */
fun debug(vararg args: Any) = genericLog(args, DEBUG)

// _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

/**
 * Write the log into the file
 */
@Synchronized
internal fun writeToFile(message: String) {
    if (printWriter != null) {
        printWriter!!.println(message)
        printWriter!!.flush()
    } else
        error("Please init Logger", CONSOLE_ONLY)
}

// _.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-._.-.

/**
 * The log's hour
 * hh:mm:ss:nnn
 */
internal fun getDate(): String {
    val now = LocalDateTime.now()

    return (now.year.toString()
            + "-" + String.format("%02d", now.monthValue)
            + "-" + String.format("%02d", now.dayOfMonth)
            + "@" + String.format("%02d", now.hour)
            + "-" + String.format("%02d", now.minute)
            + "-" + String.format("%03d", now.second))
}

/**
 * The log's date
 * yyyy-mm-dd@hh-mm-ss
 */
internal fun getHour(): String {
    val now = LocalDateTime.now()

    return (String.format("%02d", now.hour)
            + ":" + String.format("%02d", now.minute)
            + ":" + String.format("%02d", now.second)
            + ":" + String.format("%02d", now.nano / 1000))
}